backend/routes/content.py

```
@@ -710,6 +707,41 @@ def get_bonus_challenges():
+            return jsonify({"error": "Bonus challenge not found"}), 404
+
+        user_module = UserModule.query.filter_by(
+            user_id=user_id, module_id=challenge_id
+        ).first()
+
+        if user_module is None:
+            user_module = UserModule(user_id=user_id, module_id=challenge_id)
+        # deduct gems for bonus challenge purchase
+        user = User.query.get(user_id)
+        if user.gems < GEMS_FOR_BONUS_CHALLENGE:
+            logger.error(f"Insufficient gems to buy bonus challenge for user {user_id}")
```

Ruff (0.8.2)

Logging statement uses f-string

(G004)